library/dnet.c

```diff
@@ -393,6 +404,384 @@ static int dnet_iterator_callback_send(void *priv, void *data, uint64_t dsize)
+	ctl.cflags = 0;
+
+	ctl.fd = fd;
+	ctl.local_offset = data_offset;
+	ctl.cmd = DNET_CMD_WRITE;
+	ctl.ts.tv_sec = re->timestamp.tsec;
+	ctl.ts.tv_nsec = re->timestamp.tnsec;
+
+	wp = malloc(sizeof(struct dnet_iterator_server_send_write_private) + dsize);
+	if (!wp) {
+		dnet_log(n, DNET_LOG_ERROR, "%s: Interrupting iterator because failed to allocate write private data",
+				dnet_dump_id(&send->cmd.id));
+		err = -ENOMEM;
+		goto err_out_exit;
+	}
+
+	memset(wp, 0, sizeof(struct dnet_iterator_server_send_write_private));
+
+	atomic_init(&wp->refcnt, send->group_num);
+	wp->send = send;
+
+	// it is in CPU byte order, it will have to be converted to LE before sending response to client
+	memcpy(wp->data, re, dsize);
+	wp->dsize = dsize;
+
+	ctl.complete = dnet_iterator_server_send_complete;
+	ctl.priv = wp;
+
+	s = dnet_session_create(n);
+	if (!s) {
+		dnet_log(n, DNET_LOG_ERROR, "%s: Interrupting iterator because failed to create new session",
+				dnet_dump_id(&send->cmd.id));
+		err = -ENOMEM;
+		goto err_out_free;
+	}
+
+	dnet_session_set_trace_id(s, send->cmd.trace_id);
+	dnet_session_set_trace_bit(s, !!(send->cmd.flags & DNET_FLAGS_TRACE_BIT));
+
+	err = dnet_session_set_groups(s, send->groups, send->group_num);
+	if (err) {
+		dnet_log(n, DNET_LOG_ERROR, "%s: Interrupting iterator because failed to set %d groups",
+				dnet_dump_id(&send->cmd.id), send->group_num);
+		err = -ENOMEM;
+		goto err_out_session_destroy;
+	}
+
+	atomic_add(&send->bytes_pending, re->size);
+
+	dnet_log(n, DNET_LOG_NOTICE, "%s: %s: sending WRITE request, iterator response: %s, user_flags: %llx, ts: %lld.%09lld, "
+			"status: %d, size: %lld, iterated_keys: %lld/%lld",
+			__func__,
+			dnet_dump_id(&send->cmd.id), dnet_dump_id_str(re->key.id),
+			(unsigned long long)re->user_flags,
+			(unsigned long long)re->timestamp.tsec, (unsigned long long)re->timestamp.tnsec,
+			re->status, (unsigned long long)re->size,
+			(unsigned long long)re->iterated_keys, (unsigned long long)re->total_keys);
+
+	/*
+	 * After calling this function we do not own @wp anymore
+	 * if we will have to perform some actions on it, its reference counter
+	 * must be increased accordingly
+	 */
+	dnet_trans_create_send_all(s, &ctl);
+	dnet_session_destroy(s);
+
+	if (!err)
+		err = send->write_error;
+
+	return err;
+
+err_out_session_destroy:
+	dnet_session_destroy(s);
+err_out_free:
+	free(wp);
+err_out_exit:
+	dnet_server_send_put(send);
+	return err;
+}
+
+/*!
+ * Internal callback that sends result to different server in the storage as WRITE command
+ */
+static int dnet_iterator_callback_server_send(void *priv, void *data, uint64_t dsize, int fd, uint64_t data_offset)
+{
+	struct dnet_server_send_ctl *send = priv;
+	struct dnet_net_state *st = send->state;
+	struct dnet_iterator_response *re = data;
+	int err;
+
+	dnet_convert_iterator_response(re);
+
+	err = dnet_server_send_write(send, re, dsize, fd, data_offset);
```

Comment:
Verifying checksum might not that good idea actually - there might be the case that server runs with no-checksum flag, or it will run with this flag, but iterator has failed to copy data or to send it to client for recovery. This ends up with some keys have not been recovered although they can be read after no-checksum flag has been turned on.
Uncommitted keys raise different issue - send iterator writes them as completed writes, this is probably not what is expected. Data iterator for instance returns it too, but recovery will skip it, since it will read the key and it will fail.
I will add a check which skips uncommitted data and add this into documentation.

This is also a problem with current code - reading data will fail, recovery will skip the key, it will not be copied, turning off checksum verification *returns* key back.

I'm not sure whether this is a real problem or not, I would rather forbid writes without checksums.